### PR TITLE
#19080 system container

### PR DESCRIPTION
**Pull request by jdotcms**
Adding changes to handle the headless container: this is a container that does not need any content type layout, instead of it uses a generic layout for all content types or the content types selected by the user by config.

This generic layout it's the **default_container.vtl** file.

Can provide a list of content types that this container can use `$dotJSON.put("useDefaultLayout", "*")` at the **container.vtl**, as a note: `*` means all content types. This will be also respecting content type permissions based on the user when adding content.

Both vtl's were added to the empty starter.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java [hunk 1]
```diff
@@ -42,15 +42,28 @@ public FileAssetContainer() {
     @JsonIgnore
     private transient FileAsset postLoopAsset = null;
 
+    @JsonIgnore
+    private transient FileAsset defaultContainerLayoutAsset = null;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java:46]: Codacy found an issue: [Avoid using redundant field initializer for 'defaultContainerLayoutAsset'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java:46]: Codacy found an issue: [Avoid using redundant field initializer for 'defaultContainerLayoutAsset'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java [hunk 2]
```diff
@@ -56,4 +72,81 @@ public void get_container_structure_from_file_name_() throws Exception {
...
+
+        //Getting the current default host
+        final String pathRoot             = "/application/containers/test"+System.currentTimeMillis();
+        final Host currentDefaultHost     = APILocator.getHostAPI().findDefaultHost(APILocator.systemUser(), false);
+        final Folder containerFolder      = this.createAppContainerTest(currentDefaultHost, pathRoot);
+        final String contentTypeVariable  = "testfa"+System.currentTimeMillis();
+        ContentType fileAssetContentType  = ImmutableFileAssetContentType.builder()
+                .name("FileAssetTestContentType").variable(contentTypeVariable).build();
+        fileAssetContentType              = APILocator.getContentTypeAPI(APILocator.systemUser()).save(fileAssetContentType);
+        final List<FileAsset> assets      = Arrays.asList(
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java:92]: Codacy found an issue: [Avoid unused local variables such as 'assets'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 3]
```diff
@@ -116,6 +135,70 @@ public void test_getHost_demo_default_demo() throws Exception {
...
+        // Create a piece of content for the default host
+        return APILocator.getFileAssetAPI().fromContentlet(
+                APILocator.getContentletAPI().checkin(fileAsset, user, false));
+    }
+
+    private Folder createAppContainerTest(final Host currentDefaultHost) throws DotDataException, DotSecurityException {
 
+        final String pathRoot   = "/application/containers/test"+System.currentTimeMillis();
+        final Folder rootFolder = APILocator.getFolderAPI().createFolders(pathRoot, currentDefaultHost, APILocator.systemUser(), false);
+        return rootFolder;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:201]: Codacy found an issue: [Consider simply returning the value vs storing it in local variable 'rootFolder'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java [hunk 4]
```diff
@@ -56,4 +72,81 @@ public void get_container_structure_from_file_name_() throws Exception {
...
+            Assert.assertTrue("Content Type: " + contentType.variable() + " does not included on the container",
+                    containerStructures.stream().anyMatch(containerStructure -> containerStructure.getStructureId().equals(contentType.id())));
+        }
+    }
+
+    private FileAsset createFileAsset (final String fileName1,
+                                       final String fileExtension,
+                                       final String fileContent,
+                                       final Folder root1,
+                                       final ContentType fileAssetContentType) throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java:126]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java:46]: Codacy found an issue: [Avoid using redundant field initializer for 'defaultContainerLayoutAsset'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 5]
```diff
@@ -1,14 +1,33 @@
 package com.dotmarketing.portlets.containers.business;
 
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotcms.contenttype.model.type.ImmutableFileAssetContentType;
 import com.dotcms.datagen.ContainerAsFileDataGen;
+import com.dotcms.datagen.FolderDataGen;
 import com.dotcms.datagen.SiteDataGen;
+import com.dotcms.rendering.velocity.viewtools.content.FileAssetMap;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:8]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.rendering.velocity.viewtools.content.FileAssetMap'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 6]
```diff
@@ -1,14 +1,33 @@
 package com.dotmarketing.portlets.containers.business;
 
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotcms.contenttype.model.type.ImmutableFileAssetContentType;
 import com.dotcms.datagen.ContainerAsFileDataGen;
+import com.dotcms.datagen.FolderDataGen;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:6]: Codacy found an issue: [Avoid unused imports such as 'com.dotcms.datagen.FolderDataGen'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java [hunk 7]
```diff
@@ -1,6 +1,18 @@
...
 import static org.junit.Assert.assertEquals;
+
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotcms.contenttype.model.type.ImmutableFileAssetContentType;
+import com.dotmarketing.beans.ContainerStructure;
+import com.dotmarketing.beans.Host;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.portlets.containers.model.Container;
+import com.dotmarketing.portlets.containers.model.FileAssetContainer;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java:12]: Codacy found an issue: [Avoid unused imports such as 'com.dotmarketing.portlets.containers.model.FileAssetContainer'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java [hunk 8]
```diff
@@ -56,4 +72,81 @@ public void get_container_structure_from_file_name_() throws Exception {
...
+
+        // Create a piece of content for the default host
+        return APILocator.getFileAssetAPI().fromContentlet(
+                APILocator.getContentletAPI().checkin(fileAsset, APILocator.systemUser(), false));
+    }
+
+    private Folder createAppContainerTest(final Host currentDefaultHost, final String pathRoot) throws DotDataException, DotSecurityException {
+
+        final Folder rootFolder = APILocator.getFolderAPI().createFolders(pathRoot, currentDefaultHost, APILocator.systemUser(), false);
+        return rootFolder;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/ContainerStructureFinderStrategyResolverTest.java:149]: Codacy found an issue: [Consider simply returning the value vs storing it in local variable 'rootFolder'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 9]
```diff
@@ -1,14 +1,33 @@
...
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotSecurityException;
 import com.dotmarketing.portlets.ContentletBaseTest;
+import com.dotmarketing.portlets.containers.model.Container;
 import com.dotmarketing.portlets.containers.model.FileAssetContainer;
+import com.dotmarketing.portlets.contentlet.model.IndexPolicy;
+import com.dotmarketing.portlets.fileassets.business.FileAsset;
+import com.dotmarketing.portlets.fileassets.business.FileAssetAPI;
+import com.dotmarketing.portlets.folders.model.Folder;
+import com.liferay.portal.model.User;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:20]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.portal.model.User'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 10]
```diff
@@ -116,6 +135,70 @@ public void test_getHost_demo_default_demo() throws Exception {
...
+        Assert.assertEquals(1   , FileAssetContainer.class.cast(container).getContainerStructuresAssets().size());
+        Assert.assertNotNull(FileAssetContainer.class.cast(container).getDefaultContainerLayoutAsset());
+        Assert.assertEquals("default_container.vtl",FileAssetContainer.class.cast(container).getDefaultContainerLayoutAsset().getFileName());
+    }
+
+    private FileAsset createFileAsset (final String fileName1,
+                                       final String fileExtension,
+                                       final String fileContent,
+                                       final Folder root1,
+                                       final ContentType fileAssetContentType) throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:177]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 11]
```diff
@@ -1,14 +1,33 @@
...
+import com.dotmarketing.portlets.fileassets.business.FileAsset;
+import com.dotmarketing.portlets.fileassets.business.FileAssetAPI;
+import com.dotmarketing.portlets.folders.model.Folder;
+import com.liferay.portal.model.User;
+import com.liferay.util.FileUtil;
 import org.junit.Assert;
 import org.junit.Test;
 
+import java.io.File;
+import java.io.IOException;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:26]: Codacy found an issue: [Avoid unused imports such as 'java.io.IOException'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java [hunk 12]
```diff
@@ -1,14 +1,33 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/containers/business/FileAssetContainerUtilTest.java:29]: Codacy found an issue: [Avoid unused imports such as 'java.util.Map'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6347242)

**Comment by erickgonzalez** [comment]
Let's hold this one until `release-21.03` is merged into master since on `release-21.03` we updated the empty starter as well.

**Review by freddyucv — APPROVED** [review]
Are you sure that you are covered all the case in the test?

**Comment by erickgonzalez** [comment]
> Are you sure that you are covered all the case in the test?

All that we could think of :)

**Review by erickgonzalez — APPROVED** [review]

**Comment by erickgonzalez** [comment]
A new empty starter has been created and uploaded with the following changes:
* add getting started portlet.
* update portlets icons since font awesome was removed.
* add system container directory and files.
https://github.com/dotCMS/core/pull/19456/commits/80a8fdce91cee22dfa4c95658f409f7c7b5392b2

**Review by nollymar — APPROVED** [review]

**Pull request merged**